include/deal.II/optimization/solver_bfgs.h

```diff
@@ -0,0 +1,402 @@
+
+
+protected:
+  /**
+   * Additional data to the solver.
+   */
+  const AdditionalData additional_data;
+
+  /**
+   * Signal used to perform line search.
+   */
+  boost::signals2::signal<
+    Number(Number &f, VectorType &x, VectorType &g, const VectorType &p)>
+    line_search_signal;
+
+  /**
+   * Signal used to perform preconditioning.
+   */
+  boost::signals2::signal<void(VectorType &                         g,
+                               const FiniteSizeHistory<VectorType> &s,
+                               const FiniteSizeHistory<VectorType> &y)>
+    preconditioner_signal;
+};
+
+
+// -------------------  inline and template functions ----------------
+#ifndef DOXYGEN
+
+template <typename VectorType>
+SolverBFGS<VectorType>::AdditionalData::AdditionalData(
+  const unsigned int max_history_size_,
+  const bool         debug_output_)
+  : max_history_size(max_history_size_)
+  , debug_output(debug_output_)
+{}
+
+
+
+template <typename VectorType>
+SolverBFGS<VectorType>::SolverBFGS(SolverControl &       solver_control,
+                                   const AdditionalData &data)
+  : Solver<VectorType>(solver_control)
+  , additional_data(data)
+{}
+
+
+
+template <class VectorType>
+boost::signals2::connection
+SolverBFGS<VectorType>::connect_line_search_slot(
+  const std::function<
+    Number(Number &f, VectorType &x, VectorType &g, const VectorType &p)> &slot)
+{
+  Assert(line_search_signal.empty(),
+         ExcMessage("One should not attach more than one line search signal."));
+  return line_search_signal.connect(slot);
+}
+
+
+
+template <class VectorType>
+boost::signals2::connection
+SolverBFGS<VectorType>::connect_preconditioner_slot(
+  const std::function<void(VectorType &                         g,
+                           const FiniteSizeHistory<VectorType> &s,
+                           const FiniteSizeHistory<VectorType> &y)> &slot)
+{
+  Assert(preconditioner_signal.empty(),
+         ExcMessage(
+           "One should not attach more than one preconditioner signal."));
+  return preconditioner_signal.connect(slot);
+}
+
+
+
+template <typename VectorType>
+void
+SolverBFGS<VectorType>::solve(
+  const std::function<typename VectorType::value_type(VectorType &x,
+                                                      VectorType &f)> &compute,
+  VectorType &                                                         x)
+{
+  // Also see scipy Fortran implementation
+  // https://github.com/scipy/scipy/blob/master/scipy/optimize/lbfgsb_src/lbfgsb.f
+  // and Octave-optim implementation:
+  // https://sourceforge.net/p/octave/optim/ci/default/tree/src/__bfgsmin.cc
+  LogStream::Prefix prefix("BFGS");
+
+  // default line search:
+  bool       first_step = true;
+  Number     f_prev     = 0.;
+  VectorType x0;
+  const auto default_line_min =
+    [&](Number &f, VectorType &x, VectorType &g, const VectorType &p) {
+      const Number f0 = f;
```

Comment:
You need this lambda only in the `if` block following this, where you check whether a line search function had been provided. So move the declaration of the lambda into that block.